lib/migration.c

```diff
@@ -98,77 +98,81 @@ init_migration(const vfu_migration_callbacks_t * callbacks,
-    }
-
-    if (!vfio_migr_state_transition_is_valid(migr->info.device_state,
-                                              *device_state)) {
-        vfu_log(vfu_ctx, LOG_ERR, "bad transition from state %s to state %s",
-               migr_states[migr->info.device_state].name,
-               migr_states[*device_state].name);
-        return ERROR_INT(EINVAL);
-    }
-
-    switch (*device_state) {
+vfu_migr_state_t
+MOCK_DEFINE(migr_state_vfio_to_vfu)(uint32_t device_state)
+{
+    switch (device_state) {
         case VFIO_DEVICE_STATE_STOP:
-            ret = migr->callbacks.transition(vfu_ctx, VFU_MIGR_STATE_STOP);
-            break;
+            return VFU_MIGR_STATE_STOP;
         case VFIO_DEVICE_STATE_RUNNING:
-            ret = migr->callbacks.transition(vfu_ctx, VFU_MIGR_STATE_RUNNING);
-            break;
+            return VFU_MIGR_STATE_RUNNING;
         case VFIO_DEVICE_STATE_SAVING:
             /*
              * FIXME How should the device operate during the stop-and-copy
              * phase? Should we only allow the migration data to be read from
              * the migration region? E.g. Access to any other region should be
              * failed? This might be a good question to send to LKML.
              */
-            ret = migr->callbacks.transition(vfu_ctx,
-                                             VFU_MIGR_STATE_STOP_AND_COPY);
-            break;
+            return VFU_MIGR_STATE_STOP_AND_COPY;
         case VFIO_DEVICE_STATE_RUNNING | VFIO_DEVICE_STATE_SAVING:
-            ret = migr->callbacks.transition(vfu_ctx, VFU_MIGR_STATE_PRE_COPY);
-            break;
+            return VFU_MIGR_STATE_PRE_COPY;
         case VFIO_DEVICE_STATE_RESUMING:
-            ret = migr->callbacks.transition(vfu_ctx, VFU_MIGR_STATE_RESUME);
-            break;
-        default:
-            assert(false);
+            return VFU_MIGR_STATE_RESUME;
     }
+    return -1;
+}
+
+int
+MOCK_DEFINE(state_trans_notify)(vfu_ctx_t *vfu_ctx,
+                                 int (*fn)(vfu_ctx_t*, vfu_migr_state_t),
+                                 uint32_t vfio_device_state)
+{
+    /*
+     * We've already checked that device_state is valid by calling
+     * vfio_migr_state_transition_is_valid.
+     */
+    return fn(vfu_ctx, migr_state_vfio_to_vfu(vfio_device_state));
+}
 
-    if (ret == 0) {
-        migr->info.device_state = *device_state;
-        migr_state_transition(migr, VFIO_USER_MIGR_ITER_STATE_INITIAL);
-    } else if (ret < 0) {
-        vfu_log(vfu_ctx, LOG_ERR, "failed to transition to state %d: %m",
-                *device_state);
+ssize_t
+MOCK_DEFINE(migr_trans_to_valid_state)(vfu_ctx_t *vfu_ctx, struct migration *migr,
+                                       uint32_t device_state, bool notify)
+{
+    if (notify) {
+        int ret = state_trans_notify(vfu_ctx, migr->callbacks.transition,
+                                     device_state);
+        if (ret < 0) {
+            return ERROR_INT(-ret);
```

Comment:
why the -ret?

Reply:
`state_trans_notify` returns whatever the user callback returns, which according to our documentation must return `-errno` on error.

Reply:
Did I miss somewhere in the docs that still says this? They are supposed to have all been moved to setting -1 plus errno

Reply:
Oh you're right.